### PR TITLE
[Diffusion][MAGI-2] Fuse multi-head MoE route construction

### PR DESCRIPTION
**Pull request by IDEA-V**
## Purpose
 #7085 M1

## Summary
MAGI-2 routes every one of twelve 256-wide hidden-state heads into its own 256-expert bank, so route construction runs on a [12, tokens, 256] fp32 score tensor 36 times per denoise step.  The unfused path walked that tensor once per routing stage and spent most of its time in torch.topk, which is heavily over-provisioned for selecting 6 of 256: on a B200 at 540p shapes (29184 tokens) topk alone cost 4.62 ms against a 0.07 ms read of the same logits.

Fuse sigmoid, the selection bias, top-k, the probability gather and the L1 normalization into one Triton kernel that keeps a whole expert bank in registers and reads the logits exactly once.  The winner of each of the six top-k rounds is recovered with a max plus a min-index reduction rather than tl.argmax, which is several times more expensive at this shape, and the score is evaluated with libdevice's exp so the fused sigmoid stays within one ULP of torch.sigmoid instead of the 29 ULP that tl.exp's ex2 approximation gives.

Rebuild the CSR layout the same way it was, only cheaper: sort the flattened expert keys as int16 while they fit, derive token ids from the permutation instead of gathering them out of a materialized index tensor, and take the offsets from a binary search over the sorted keys rather than torch.bincount, which synchronizes.  The layout is bit-identical to the reference.

## Performance

Measured on one NVIDIA B200 with 12 heads, 256 experts per head, and top-6 routing:

| Tokens | Reference | Fused | Speedup | Workspace |
|---:|---:|---:|---:|---:|
| 1,024 | 416.8 µs | 143.5 µs | 2.90× | 25.2 → 3.0 MiB |
| 7,674 | 1,502.3 µs | 191.4 µs | 7.85× | 188.8 → 23.0 MiB |
| 29,184 | 5,020.4 µs | 451.7 µs | 11.11× | 717.4 → 84.5 MiB |
| 58,368 | 9,817.7 µs | 840.4 µs | 11.68× | 1,434.8 → 169.1 MiB |

  ## Test Plan

  vLLM Version: 0.28.0

  Tested on one NVIDIA B200 using Python 3.11 and PyTorch 2.13.0+cu130.

  python -m pytest tests/config/test_environment_variables.py -q
  python -m pytest tests/diffusion/models/magi2/test_mh_moe_routing.py -q

  ## Test Result

  ```text
  tests/config/test_environment_variables.py:
  13 passed

  tests/diffusion/models/magi2/test_mh_moe_routing.py:
  33 passed

  All tests passed with no failures or skips. The routing suite covers:

  - Fused/reference output parity across representative shapes
  - Routing with and without expert bias
  - Route normalization enabled and disabled
  - CSR layout equivalence and ordering invariants
  - End-to-end expert output parity
  - CPU and unsupported-input fallback behavior
  - Invalid-input validation
  - A relative routing performance regression threshold of at least 3×

### CODE REVIEW

File: vllm_omni/diffusion/models/magi2/mh_moe.py [hunk 1]
```diff
@@ -77,15 +76,226 @@ def compute_topk_probs_and_indices(
...
+    # fp32 only: the reference evaluates sigmoid and the L1 norm in the logit
+    # dtype, and reproducing narrow-dtype rounding is not worth a second kernel.
+    if router_logits.dtype != torch.float32:
+        return False
+    if router_logits.stride(-1) != 1 or router_logits.shape[1] == 0:
+        return False
+    if triton.next_power_of_2(router_logits.shape[-1]) > _MAX_FUSED_EXPERTS_PAD:
+        return False
+    if expert_bias is not None and (expert_bias.dtype != torch.float32 or not expert_bias.is_contiguous()):
+        return False
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/magi2/mh_moe.py]: [P2] Validate bias size before selecting the fused path

The reference supports a per-head scalar bias shaped `[H,1]` through broadcasting, but this guard accepts it while the kernel reads `H*num_experts` elements. With logits `[8,7,...,1]` and a contiguous `[1,1]` zero-bias view, I reproduced reference indices `[0,1,2]` versus fused `[7,6,5]`, determined by storage outside the view. Require the expected bias element count or fall back for broadcast shapes.

File: vllm_omni/diffusion/models/magi2/mh_moe.py [hunk 2]
```diff
@@ -77,15 +76,226 @@ def compute_topk_probs_and_indices(
...
+        # shape, so recover the winner with a second reduction.  Ties resolve to
+        # the lowest expert id, which torch.topk leaves unspecified.
+        best_expert = tl.min(tl.where(selection_scores == best_score[:, None], expert_offsets, experts_pad), axis=1)
+        selected = expert_offsets[None, :] == best_expert[:, None]
+        # The bias steers selection only; the route weight is the unbiased score.
+        probability = tl.sum(tl.where(selected, router_scores, 0.0), axis=1)
+        is_route = route_offsets == route
+        topk_probs += tl.where(is_route[None, :], probability[:, None], 0.0)
+        topk_indices += tl.where(is_route[None, :], best_expert[:, None].to(tl.int32), 0)
+        selection_scores = tl.where(selected, float("-inf"), selection_scores)
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/magi2/mh_moe.py:169]: [P2] Exclude previously selected experts independently of score

Setting selected scores to `-inf` does not exclude them once the remaining maximum is also `-inf`. With zero logits, `top_k=3`, and bias `-inf` everywhere except expert 2, CUDA returns `[2,0,0]`, whereas the reference returns three distinct experts. This changes the weighted expert output beyond an unspecified tie order. Track eligibility separately so already-selected experts cannot win subsequent rounds.

**Review by hsliuustc0106 — COMMENTED** [review]
Found two edge-case correctness issues, detailed inline.

Validation: 32 CUDA parity combinations passed, including CSR equality; both issues reproduced on GPU using an isolated routing harness with PyTorch 2.11.0+cu130 and Triton 3.6.0. Full pytest execution was unavailable because the GPU environment lacks pytest.

File: vllm_omni/diffusion/models/magi2/mh_moe.py (hunk 1 above)
Comment (IDEA-V) [reply vllm_omni/diffusion/models/magi2/mh_moe.py]: Thanks for the precise repro. The guard checked the bias dtype and contiguity but never its element count, while the kernel indexes a dense `[heads, num_experts]` bank.

Rather than add a size check and fall back, the broadcast is now materialized, so a per-head scalar bias keeps the fused path:

```python
def _dense_expert_bias(expert_bias, heads, num_experts):
    dense = expert_bias.view(heads, -1).expand(heads, num_experts)
    return dense.to(torch.float32).contiguous()
```

The trailing `contiguous()` is load bearing — `expand` leaves a zero stride the kernel cannot see. Shapes the reference's own broadcast would reject now raise from `view` / `expand` instead of reaching the kernel, and every step is a no-op for a dense fp32 bias (verified by `data_ptr()` identity).

I also dropped the fp32 requirement on the bias: the reference adds a narrow bias to fp32 scores and gets fp32 by promotion, so upcasting here is bit-identical and need not cost the fused path. The guard keeps only contiguity, which is what `view` actually requires.

New tests: `test_broadcast_expert_bias_matches_reference` (`[H,1]`, `[H]`, `[H,1,1]`) and `test_narrow_dtype_expert_bias_matches_reference` (bf16, fp16).

File: vllm_omni/diffusion/models/magi2/mh_moe.py (hunk 2 above)
Comment (IDEA-V) [reply vllm_omni/diffusion/models/magi2/mh_moe.py:169]: Thanks for the finding.

I fixed the ambiguity from the other side instead of tracking eligibility separately, because I measured that version and it was expensive. An unbiased sigmoid is always inside `(0, 1)`, so the bias is the only way a score reaches `-inf`; clamping it makes `-inf` unreachable for anything selectable and restores it as a private retirement sentinel, leaving the loop body unchanged:

```python
selection_scores = router_scores + tl.maximum(bias, _MIN_FINITE_FP32)[None, :]
```

Dead and padded lanes keep `-inf` and must never be selectable anyway, and `top_k <= num_experts` guarantees an unretired live lane every round whose score is `> -inf`, so no `-inf` lane can ever match `best_score`.

Overhead vs. the current fused kernel, on one A100-80GB (torch 2.13.0+cu130, Triton 3.7.1), 12 heads x 256 experts, top-6:

| tokens | separate eligibility mask | clamp (chosen) |
|---:|---:|---:|
| 1,024 | +29% | +20% |
| 7,674 | +36% | +7.5% |
| 29,184 | +13% | +3.2% |
| 58,368 | +15% | +2.1% |

Behavior is unchanged except in the corner your repro exercises: a bias of exactly `-inf` is treated as `-3.4e38`, which still sorts below every finite score. When a whole row is masked that way `torch.topk` leaves the order unspecified; the fused path returns the lowest expert ids. Both return `top_k` distinct experts, which was the defect.

New test: `test_exhausted_selection_scores_still_pick_distinct_experts`, which is your repro verbatim — zero logits, `top_k=3`, `-inf` bias everywhere but expert 2. It asserts three distinct experts, and fails with `[2, 0, 0]` against the kernel as it stands on this PR.

**Comment by hsliuustc0106** [comment]
Local routing A/B results for head `c12bb67b6d98108ef82d5c6feaadc2d74fc68d40` against base `eeee25791327d94bd5ea5460f5df7c1f7198d4e8`.

The candidate consistently reduced routing-plus-CSR latency in both measured rounds on an NVIDIA L20X. Extra peak allocated GPU memory decreased by approximately 88.2%.

| Tokens | Base median range (ms) | Head median range (ms) | Paired speedup range | Extra peak MiB, base → head |
|---:|---:|---:|---:|---:|
| 1,024 | 0.4294–0.4316 | 0.1700–0.1741 | 2.48–2.53× | 25.17 → 2.97 |
| 7,674 | 1.6415–1.6459 | 0.1876–0.1880 | 8.73–8.77× | 188.78 → 22.23 |
| 29,184 | 5.5357–5.5421 | 0.5313–0.5324 | 10.40–10.43× | 717.40 → 84.53 |
| 58,368 | 10.7806–10.7818 | 0.9877–0.9879 | 10.91–10.92× | 1434.80 → 169.06 |

**Method:** Python 3.12.13, PyTorch 2.11.0+cu130, Triton 3.6.0, driver 570.133.20. Same scheduler-reserved physical GPU, CPU affinity 0–7, and eight Torch CPU threads throughout. FP32 logits `[12, tokens, 256]`, top-6, seed 0 reset per shape, dense FP32 bias `randn * 0.01`, normalization enabled. One CPU-only preflight and one feasibility group across all four shapes, then exactly two measured rounds per implementation: A/B followed by B/A. Each stage used 20 excluded warmups and 50 CUDA-event samples. Feasibility/checks/JIT preparation took 2181.12 ms and was excluded. No profiler, checkpoint downloads, or `torch.compile`.

The harness extracted the original routing function bodies, decorators, and required constants from the two pinned SHAs, replacing package initialization with direct Torch/Triton imports. No production source edits. Timing above directly measures `global_sort_routes(*compute_topk_probs_and_indices(...), 256)`, rather than summing independently measured component medians as in the PR table. Memory is extra peak PyTorch **allocated** memory, not total HBM or reserved memory. Different hardware, software, and the updated head mean this is not an exact reproduction of the original B200 numbers.

**Correctness:** All candidate weights were checked against base sigmoid/gather/normalize at the candidate-selected expert IDs with `rtol=atol=1e-6`; maximum absolute error was `5.96e-8`. Selected biased scores differed from base top-k by at most `5.96e-8` (guard: `2e-7`). Expert IDs were unique. CSR payloads were bit-identical when given identical routes.

| Tokens | Compared rows | Ordered-route mismatch rows | Expert-set mismatch rows |
|---:|---:|---:|---:|
| 1,024 | 12,288 | 0 | 0 |
| 7,674 | 92,088 | 1 | 0 |
| 29,184 | 350,208 | 4 | 1 |
| 58,368 | 700,416 | 10 | 1 |

This passes the declared local numerical guards, **but is not exact routing parity**: near-equal or tied scores can choose different experts, and similar scores do not bound differences in expert outputs. Full-checkpoint T2VA/I2VA quality, end-to-end serving benefit, GPU expert execution, and distributed/compile/DLO/Cache-DiT combinations were not validated by this experiment.

<details>
<summary>Per-round combined-call timing spread (ms)</summary>

| Tokens | Round | Implementation | p10 | Median | p90 |
|---:|---:|---|---:|---:|---:|
| 1,024 | 1 | base | 0.42803 | 0.43158 | 0.43629 |
| 1,024 | 1 | head | 0.17024 | 0.17410 | 0.18406 |
| 7,674 | 1 | base | 1.63594 | 1.64587 | 1.65632 |
| 7,674 | 1 | head | 0.18502 | 0.18763 | 0.18960 |
| 29,184 | 1 | base | 5.53018 | 5.54208 | 5.55162 |
| 29,184 | 1 | head | 0.53066 | 0.53130 | 0.53242 |
| 58,368 | 1 | base | 10.77286 | 10.78179 | 10.78934 |
| 58,368 | 1 | head | 0.98653 | 0.98773 | 0.98870 |
| 1,024 | 2 | head | 0.16672 | 0.17003 | 0.17658 |
| 1,024 | 2 | base | 0.42669 | 0.42942 | 0.43635 |
| 7,674 | 2 | head | 0.18470 | 0.18800 | 0.18954 |
| 7,674 | 2 | base | 1.63706 | 1.64146 | 1.64867 |
| 29,184 | 2 | head | 0.53139 | 0.53237 | 0.53331 |
| 29,184 | 2 | base | 5.53005 | 5.53574 | 5.54314 |
| 58,368 | 2 | head | 0.98685 | 0.98794 | 0.98909 |
| 58,368 | 2 | base | 10.77594 | 10.78059 | 10.78960 |

</details>

Raw sample arrays, source snapshots/hashes, protocol, harness, and scheduler logs are retained locally. GPU allocations and the reservation were released after completion.

File: vllm_omni/diffusion/models/magi2/mh_moe.py (hunk 1 above)
Comment (hsliuustc0106) [reply vllm_omni/diffusion/models/magi2/mh_moe.py]: Fixed — `_dense_expert_bias` now materializes every reference-legal broadcast shape (and upcasts narrow dtypes exactly like the reference's own promotion), and relaxing the guard to contiguity-only keeps the fused path instead of silently dropping it.

File: vllm_omni/diffusion/models/magi2/mh_moe.py (hunk 2 above)
Comment (hsliuustc0106) [reply vllm_omni/diffusion/models/magi2/mh_moe.py:169]: Fixed — clamping the user bias to the largest finite negative keeps a retired expert (true `-inf`) strictly unreachable, and the new exhausted-bank test pins the deterministic id-order tie-break.

**Review by hsliuustc0106 — APPROVED** [review]
Both findings are addressed with targeted kernel changes and regression tests — looks good now. (DCO/build/pre-commit green on c12bb67b; the new routing tests are B200/H100/L4-gated as before.)

**Comment by hsliuustc0106** [comment]
fix conflicts

**Comment by hsliuustc0106** [comment]
Re-verified the rebased head `2c243e795`.

The conflict was #7239 landing on main with the same import fix; the resolution folds this branch's commit into that pattern — the routing sentinel is now `HAS_TRITON`-guarded like the SwiGLU7 constants, and the kernel-local constexpr variant is gone. Both earlier fixes (`_dense_expert_bias` materialization, the `tl.maximum(bias, _MIN_FINITE_FP32)` retirement clamp) survived the rebase intact, the placeholder-import subprocess test is retained and now also stubs `triton_utils.HAS_TRITON = False` so it exercises the exact guard the module reads, and the CUDA-only speedup floor with the MI300 2.16x rationale is unchanged.

Build/pre-commit/DCO are green on this head and the branch is mergeable. Nothing further from me.
